docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,229 @@
+
+Transitions:
+
+1. `Infant` -> `Loading`, `Downloading`.
+2. `Loading` -> `Active`, `Broken`.
+3. `Broken` is the final state.
+4. `Downloading` -> `Active`, `Broken`.
+5. `Active` -> `ShuttingDown`.
+6. `ShuttingDown` -> `Shutdown`.
+7. `Shutdown` is the final state.
+
+The transition graph has no cycles, which is a good property.
+
+Generally, we have 3 subsets of states:
+
+1. Paused - tries to load itself: `Loading`, `Downloading`.
+2. Alive - can operate: `Infant`, `Active`.
+3. Dead - cannot launch any operations, possibly already dropped: `Broken`, `ShuttingDown`, `Shutdown`.
+
+## `TenantAccessor`
+
+The first part of the chapter will be about `TenantAccessor` specifically. Later, we'll discuss `TimelineAccessor`.
+
+### Top-level interface
+
+We need a proper tree-like structure with a bare minimum of global states to use RAII, unlike carefully using `shutdown_tasks` as we do now.
+
+The idea is to have the following tenant API:
+
+```rust
+/// Gets the alive tenant from the in-memory data, then applies the specified
+/// function. Returns error if there is no tenant in memory or the tenant is not
+/// ready to launch the closure.
+pub fn with_tenant<F, O>(tenant_id: TenantId, func: F) -> anyhow::Result<O>
+where
+    F: FnOnce(TenantSyncRef) -> O {}
+
+/// Gets the alive tenant from the in-memory data, then applies the specified
+/// asynchronous function. Returns error if there is no tenant in memory or
+/// tenant is not ready to launch the closure.
+pub async fn with_tenant_async<F, T, O>(tenant_id: TenantId, func: F) -> anyhow::Result<O>
+where
+    F: FnOnce(TenantAsyncRef) -> T,
+    T: Future<Output = O> {}
+```
+
+So, _we won't have access to the tenant directly_. Instead, we only use transaction-inspired API with some closure that can access the tenant. These functions will be referred to as closures.
+
+- `TenantSyncRef` is `&Tenant`, but with a limited interface not to allow simple mistakes like waiting for something inside.
+- `TenantAsyncRef` is `Arc<Tenant>` only because of current Rust's limitations of async closures. Later, it will also be a `&Tenant`. It also has a limited interface to prevent some kinds of mistakes.
+
+_These closures **are not supposed to run any long-running code**, and should be checked on code reviews._
+
+To solve a problem, we'll have a global private mapping from `TenantId` to `TenantAccessor`, accessible only by `with_tenant`. `TenantAccessor` is a kind of guard similar to `Arc` in Rust but not the same. It allows running the closures with the tenant in scope but won't run them if it's not in the `Active` state, and this accessor is responsible for all cancellations.
+
+The interface:
+
+```rust
+impl TenantAccessor {
+    /// Creates a new accessor. Check the state is not `TenantState::Infant`.
+    /// Spawns a closure in tokio runtime to wait for the state to become dead.
+    /// Then, the tenant is shut down.
+    pub fn new(tenant: Tenant) -> Self;
```

Comment:
Do not understand how this composes with `LoopedController`. Looped controller needs to have reference to tenant. Or tenant references looped controllers for gc and compaction? If tenant references them how do we get tenant properties inside gc? 

I think this RFC would benefit from sketching the bigger picture with pseudocode. Tenant with stub methods for gc/compaction which just `println!("gc")` inside. (or maybe use some tenant attribute)

Reply:
`LoopedController` does not need any tenant references, it even doesn't know about our runtime. `LoopedController ` is just some kind of SPSC channel that provides synchronisation for some looped task and its controller, sending `LoopedTaskControlFlow` enum instances. But it's not just some SPSC channel, since we can wait for control message/termination from Receiver/Sender side.

Reply:
I thought it's even more similar to concurrent `Deque<LoopedTaskControlFlow>` with a small set of rules (no pushes after `Break`, no consecutive repetitions, no more than 2 objects in deque excluding `Break`) + `DropWatcher`.